api/apps/api/src/modules/clone/export/adapters/resource-pieces.adapter.ts

```diff
@@ -1,45 +1,42 @@
-import { v4 } from 'uuid';
-import {
-  ClonePiece,
-  ComponentId,
-  ResourceId,
-  ResourceKind,
-} from '@marxan/cloning/domain';
-import { Injectable } from '@nestjs/common';
-
-import { ExportComponentSnapshot } from '../domain';
+import { DiscoveryService } from '@golevelup/nestjs-discovery';
+import { ResourceId, ResourceKind } from '@marxan/cloning/domain';
+import { Injectable, SetMetadata } from '@nestjs/common';
 import { ResourcePieces } from '../application/resource-pieces.port';
+import { ExportComponentSnapshot } from '../domain';
+
+export const ResourcePiecesProviderMetadata = Symbol(
+  `Resource pieces provider`,
+);
+
+export const ResourcePiecesProvider = (kind: ResourceKind) =>
+  SetMetadata(ResourcePiecesProviderMetadata, kind);
 
 @Injectable()
 export class ResourcePiecesAdapter implements ResourcePieces {
-  /**
-   * most likely should be a kind of facade with strategy pattern
-   *
-   * so that we can separate resolvers for Project/Scenarios
-   */
+  constructor(private readonly discovery: DiscoveryService) {}
+
+  private async getAdapterFor(kind: ResourceKind): Promise<ResourcePieces> {
+    const resourcePiecesProviders = await this.discovery.providersWithMetaAtKey<ResourceKind>(
+      ResourcePiecesProviderMetadata,
+    );
+
+    const resourcePiecesProvider = resourcePiecesProviders.find(
+      (provider) => provider.meta === kind,
+    );
+
+    if (!resourcePiecesProvider) {
+      throw new Error(`ResourcePieces adapter not found for ${kind}`);
```

Comment:
just a note - when we get to throw this error it should normally be because of a bug (or missing implementation), not something we fail to do because of user input/data available/etc. I wonder if we should make this more explicit somehow (in the Marxan bots code I used something like `No feature found with given name: this is probably a bug in this bot` as an error message.
In this case I'm not overly concerned because this error message would likely not make it through to the user if they use the web app, but it may be available for us devs (depending on how it's handled - likely turned into a 500 by the global exception filter). Just food for thought 😄

Reply:
I've changed the message to `ResourcePieces adapter not found for ${kind} resource kind. This is probably caused by a missing provider in export-adapters module` 👍